File: gymTracker/utils/Utility.m
#import "Utility.h"
#import "AppDelegate.h"

@implementation Utility

+ (void) showAlert:(NSString *)title message:(NSString *)msg
{
    UIAlertView *alert = [[UIAlertView alloc] initWithTitle:title message:msg delegate:self cancelButtonTitle:nil otherButtonTitles:@"Ok", nil];

    [alert show];
}

+ (Utility *) sharedInstance
{
    static Utility *_sharedClient = nil;
    
    static dispatch_once_t oncePredicate;
    dispatch_once(&oncePredicate, ^{
        _sharedClient = [self new];
        
        _sharedClient.dateOnlyDateFormat = [NSDateFormatter new];
        _sharedClient.dateOnlyDateFormat.dateFormat = @"dd";
        
        _sharedClient.dbDateFormat = [NSDateFormatter new];
        _sharedClient.dbDateFormat.dateFormat = @"yyyy-MM-dd";
        
        _sharedClient.userFriendlyDateFormat = [NSDateFormatter new];
        _sharedClient.userFriendlyDateFormat.dateFormat = @"MMM/dd/yyyy";
    });
    
    return _sharedClient;
}

+ (NSDate *)dateAtBeginningOfDayForDate:(NSDate *)inputDate
{
    NSCalendar *calendar = [NSCalendar currentCalendar];
    NSTimeZone *timeZone = [NSTimeZone systemTimeZone];
    [calendar setTimeZone:timeZone];
    
    NSDateComponents *dateComps = [calendar components:NSYearCalendarUnit | NSMonthCalendarUnit | NSDayCalendarUnit fromDate:inputDate];
    
    [dateComps setHour:0];
    [dateComps setMinute:0];
    [dateComps setSecond:0];
    
    NSDate *beginningOfDay = [calendar dateFromComponents:dateComps];
    return beginningOfDay;
}

+ (UIImage *)scaleImage:(UIImage *)image toSize:(CGSize)newSize
{
    float width = newSize.width;
    float height = newSize.height;
    
    UIGraphicsBeginImageContext(newSize);
    CGRect rect = CGRectMake(0, 0, width, height);
    
    float widthRatio = image.size.width / width;
    float heightRatio = image.size.height / height;
    float divisor = widthRatio > heightRatio ? widthRatio : heightRatio;
    
    width = image.size.width / divisor;
    height = image.size.height / divisor;
    
    rect.size.width  = width;
    rect.size.height = height;
    
    float offset = (width - height) / 2;
    if (offset > 0)
    {
        rect.origin.y = offset;
    }
    else
    {
        rect.origin.x = -offset;
    }
    
    [image drawInRect: rect];
    
    UIImage *smallImage = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return smallImage;
}

@end

File: gymTracker/services/Utility.swift
//
//  UtilitySwift.swift
//  gymTracker
//
//  Created by Shuvo on 9/19/18.
//  Copyright © 2018 Third Bit. All rights reserved.
//

import Foundation

class Utility: NSObject {
    
    var documentDir: String = ""
    var databaseName: String = ""
    var databasePath: String = ""
    var settings = Settings()
    var dbDateFormat = DateFormatter()
    var userFriendlyDateFormat = DateFormatter()
    var dateOnlyDateFormat = DateFormatter()
    var noImage: UIImage!
    var checkBoxImage: UIImage!
    var checkBoxMarkedImage: UIImage!
    var equipmentsList: NSMutableArray = []
    var measurementsList: NSMutableArray = []
    
    class func showAlert(title: String, message: String, in vc: UIViewController) {
        let alert =  UIAlertController(title: title, message: message, preferredStyle: UIAlertControllerStyle.alert)
        alert.addAction(UIAlertAction(title: "Ok", style: UIAlertActionStyle.default, handler: nil))
        vc.present(alert, animated: true, completion: nil)
    }

    static let sharedInstance:Utility = {
        let instance = Utility ()
        return instance
    } ()
    
    
    // MARK: Init
    override init() {
        dateOnlyDateFormat = DateFormatter()
        dateOnlyDateFormat.dateFormat = "dd"
        
        dbDateFormat = DateFormatter()
        dbDateFormat.dateFormat = "yyyy-MM-dd"
        
        userFriendlyDateFormat = DateFormatter()
        userFriendlyDateFormat.dateFormat = "MMM/dd/yyyy"
        
        noImage = UIImage(named: "no_image.jpg")
        
        checkBoxImage = UIImage(named: "checkBox.png")
        checkBoxMarkedImage = UIImage(named: "checkBoxMarked.png")
    }
    
    
//    static let sharedInstance: Utility = {
//        let sharedClient = Utility()
//
//        sharedClient.dateOnlyDateFormat = DateFormatter()
//        sharedClient.dateOnlyDateFormat.dateFormat = "dd"
//
//        sharedClient.dbDateFormat = DateFormatter()
//        sharedClient.dbDateFormat.dateFormat = "yyyy-MM-dd"
//
//        sharedClient.userFriendlyDateFormat = DateFormatter()
//        sharedClient.userFriendlyDateFormat.dateFormat = "MMM/dd/yyyy"
//
//        sharedClient.noImage = UIImage(named: "no_image.jpg")
//
//        sharedClient.checkBoxImage = UIImage(named: "checkBox.png")
//        sharedClient.checkBoxMarkedImage = UIImage(named: "checkBoxMarked.png")
//
//        return sharedClient
//    } ()
    
    class func scaleImage(image: UIImage, toSize newSize:CGSize) -> UIImage? {
        var width: Float = Float(newSize.width)
        var height: Float = Float(newSize.height)
        
        UIGraphicsBeginImageContext(newSize)
        var rect = CGRect(x: 0, y: 0, width: CGFloat(width), height: CGFloat(height))
        
        let widthRatio: Float = Float(image.size.width) / width
        let heightRatio: Float = Float(image.size.height) / height
        let divisor: Float = widthRatio > heightRatio ? widthRatio : heightRatio
        
        width = Float(image.size.width) / divisor
        height = Float(image.size.height) / divisor
        
        rect.size.width = CGFloat(width)
        rect.size.height = CGFloat(height)
        
        let offset: Float = (width - height) / 2
        if offset > 0 {
            rect.origin.y = CGFloat(offset)
        } else {
            rect.origin.x = CGFloat(-offset)
        }
        
        image.draw(in: rect)
        
        let smallImage: UIImage? = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        
        return smallImage
    }
    
    
     class func dateAtBeginningOfDay(inputDate: Date?) -> Date? {
        var calendar = Calendar.current
        let timeZone = NSTimeZone.system as NSTimeZone
        calendar.timeZone = timeZone as TimeZone
        
        var dateComps: DateComponents? = nil
        if let aDate = inputDate {
            //dateComps = calendar.dateComponents([.NSYearCalendarUnit, .NSMonthCalendarUnit, .NSDayCalendarUnit], from: aDate)
            dateComps = calendar.dateComponents([.year, .month, .day], from: aDate as Date)
        }
        
        dateComps?.hour = 0
        dateComps?.minute = 0
        dateComps?.second = 0
        
        var beginningOfDay: Date? = nil
        if let aComps = dateComps {
            beginningOfDay = calendar.date(from: aComps)
        }
        return beginningOfDay
    }
    
    
}
